src/metabase/query_analysis.clj

```diff
@@ -12,29 +12,59 @@
    [metabase.query-analysis.native-query-analyzer.replacement :as nqa.replacement]
    [metabase.util :as u]
    [metabase.util.log :as log]
+   [metabase.util.queue :as queue]
    [toucan2.core :as t2]))
 
-(def ^:dynamic *parse-queries-in-test?*
-  "Normally, a native card's query is parsed on every create/update. For most tests, this is an unnecessary
-  expense. Therefore, we skip parsing while testing unless this variable is turned on.
+(set! *warn-on-reflection* true)
 
-  c.f. [[native-analysis-active?]]"
-  false)
+(def ^:private realtime-queue-capacity 1000)
 
-(defn- native-analysis-active?
-  "Should the query run? Either we're not testing or it's been explicitly turned on.
+(defonce ^:private queue (queue/bounded-transfer-queue realtime-queue-capacity {:dedupe? true}))
+
+(def ^:dynamic *analyze-execution-in-dev?*
+  "Managing a background thread in the REPL is likely to confound and infuriate, especially when we're using it to run
+  tests. For this reason we run analysis on the main thread by default."
+  ::immediate)
+
+(def ^:dynamic *analyze-execution-in-test?*
+  "A card's query is normally analyzed on every create/update. For most tests, this is an unnecessary expense, hence
+  we disable analysis by default."
+  ::disabled)
+
+(defmacro with-execution*
+  "Override the default execution mode, except in prod."
+  [execution & body]
+  `(binding [*analyze-execution-in-dev?* ~execution
+             *analyze-execution-in-test?* ~execution]
+     ~@body))
+
+(defmacro with-queued-analysis
+  "Override the default execution mode to always use the queue. Does nothing in prod - only use this in tests."
+  [& body]
+  `(with-execution* ::queued ~@body))
+
+(defmacro with-immediate-analysis
+  "Override the default execution mode to always use the current thread. Does nothing in prod - only use this in tests."
+  [& body]
+  `(with-execution* ::immediate ~@body))
+
+(defmacro without-analysis
+  "Override the default execution mode to always use the current thread. Does nothing in prod - only use this in tests."
+  [& body]
+  `(with-execution* ::disabled ~@body))
+
+(defn- execution []
+  (case config/run-mode
+    :prod ::queued
+    :dev  *analyze-execution-in-dev?*
+    :test *analyze-execution-in-test?*))
 
-  c.f. [[*parse-queries-in-test?*]], [[public-settings/sql-parsing-enabled]]"
-  []
-  (and (public-settings/sql-parsing-enabled)
-       (or (not config/is-test?)
-           *parse-queries-in-test?*)))
 
-(defn enabled?
+(defn enabled-type?
```

Comment:
`query-type-enabled?`?

Reply:
That sounds it bit yoda to me "query analysis query type enabled?"

If we want to be more explicit how about `query-analysis/enabled-for-query-type?`?